Fix formatting and add test for dump and load binary tables

cmarmo opened the pull request:
### Description
This pull request is to address the format issue specified in #7015.
It fixes #7015.

I have add a test that failed in main and was meant to work in this pull request.
Unfortunately it seems to me that other hidden issues rise in the functions that convert a `ColDefs` object in a table header.

The test checking the table headers is still failing for three reasons:
- [x] standard comments are different from the one in the example table `astropy/io/fits/tests/data/tb.fits` : Fixed comparing only values of cards
- [x] TFORMx default formatting is different from the one used in the example file: Fixed skipping the comparison of this particular keyword
- [x] For some reason TDISPx, TNULLx, TZEROx, TSCALEx are not converted from the `ColDefs` object: Fixed skipping the comparisong of those keywords

~I was planning to:~
~- [ ] change the test in order to avoid the comparison of the comment~
~- [ ] compare the equivalence of the TFORMx value, not the exact string ('1J' ~ 'J')~
~- [ ] take a deeper dive in `columns.py`~

Before this I would be grateful if some of the maintainers could check that I am not on a wrong path.
Thanks for your understanding and for your help.
 
### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pllim commented:
Thank you! Given that you mentioned there are things need to be done on this PR still, I have marked this as draft for now. 😸 

Probably needs a change log too.

saimn commented on astropy/io/fits/hdu/table.py:
```diff
@@ -1294,7 +1294,7 @@ def _dump_coldefs(self, fileobj):
             line = [column.name, column.format]
             attrs = ['disp', 'unit', 'dim', 'null', 'bscale', 'bzero']
-            line += ['{:16s}'.format(value if value else '""')
+            line += [f'{value}' if value else '""'
```
You should keep the `16s`, so something like this :
```suggestion
            line += ['{!s:16s}'.format(value if value else '""')
```

saimn commented:
Thanks for working on this @cmarmo ! 
For the header comparison, we should probably avoid comparing the structural keywords (i.e. the `T*` ones that are updated automatically when setting the data). So maybe we just check the history card is dumped and reloaded.
Also I don't remember exactly but I think the original was about adding similar tests for `fits.tabledump` and `fits.tableload`.

cmarmo commented on astropy/io/fits/hdu/table.py:
```diff
@@ -1294,7 +1294,7 @@ def _dump_coldefs(self, fileobj):
             line = [column.name, column.format]
             attrs = ['disp', 'unit', 'dim', 'null', 'bscale', 'bzero']
-            line += ['{:16s}'.format(value if value else '""')
+            line += [f'{str(value):16s}' if value else '""'
```
Do you mind if I keep the 'f-string' formatting... If I understand correctly no python older than 3.6 is supported by astropy... right?

saimn replied:
The problem is that the empty string case should also be padded, and I think would be clearer with `.format` in this case (you could put the if/else in the fstring but that would make multiple levels of quote which is not ideal).

cmarmo commented:
@saimn , thanks for your answer!

> For the header comparison, we should probably avoid comparing the structural keywords (i.e. the `T*` ones that are updated automatically when setting the data). So maybe we just check the history card is dumped and reloaded.

I understood your comment. The current version of the test checks for the values of the header keywords and failed. I am ok to only check the history, but I would be grateful if you could clarify if the differences in the keyword values are expected. Also, if the TNULL keyword is in the `ColDefs` object, shouldn't be in the loaded header too?
Thanks for your patience. 

> Also I don't remember exactly but I think the original was about adding similar tests for `fits.tabledump` and `fits.tableload`.

I have added a test for the dump function, I'm not sure I am testing the right thing though...
Should I add a test for the load function? It will probably test the same thing than `test_dump_load_round_trip`....
Thanks for your patience, again.

saimn commented:
I'm not sure what's best for the header comparison, I will look more in detail.

> I have added a test for the dump function, I'm not sure I am testing the right thing though...
Should I add a test for the load function? It will probably test the same thing than test_dump_load_round_trip....

I was speaking the convenience functions, `fits.tabledump` and `fits.tableload`, which are shortcuts to opening the file and using `BinTableHDU.dump`. `BinTableHDU.dump/load` is tested here (not sure if all lines are covered though), so what should be tested is the convenience functions, with almost the same test.

cmarmo commented:
> I was speaking the convenience functions, `fits.tabledump` and `fits.tableload`, which are shortcuts to opening the file and using `BinTableHDU.dump`. `BinTableHDU.dump/load` is tested here (not sure if all lines are covered though), so what should be tested is the convenience functions, with almost the same test.

Right, sorry I didn't realize before. I will do.

pep8speaks commented:
Hello @cmarmo :wave:! It looks like you've made some changes in your pull request, so I've checked the code again for style.




* In the file [`astropy/io/fits/tests/test_convenience.py`](https://github.com/astropy/astropy/blob/b360ce301257f363e0f30d160a2b182841d13bc0/astropy/io/fits/tests/test_convenience.py):

> [Line 281:43](https://github.com/astropy/astropy/blob/b360ce301257f363e0f30d160a2b182841d13bc0/astropy/io/fits/tests/test_convenience.py#L281): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

cmarmo commented:
@saimn , thanks for your patience, I think it's better now.

saimn commented:
Hi @cmarmo,
Thanks, it seems almost ready, though there is still a test failure in the header comparison:
```
>               assert card.value == new_card.value
E               AssertionError: assert '1J' == 'J'
```
Could you also add a changelog entry ?

cmarmo commented:
> Thanks, it seems almost ready, though there is still a test failure in the header comparison:

Indeed, this is the issue I mentioned in the [description](https://github.com/astropy/astropy/pull/12526#issue-1061759263).
I'm not happy about how I fixed it... skipping the comparison of the problematic cards.
I have added a comment: it is not clear to me if those keywords issues are expected or not.
In the meanwhile the test passes.

saimn commented on astropy/io/fits/tests/test_convenience.py:
```diff
@@ -255,6 +270,41 @@ def test_tabledump(self):
+                # Skip unloaded keywords (should this be fixed?)
+                if card.keyword[0:5] not in ['TFORM', 'TDISP', 'TNULL', 'TZERO', 'TSCAL']:
+                    new_card = new_tbhdu.header.cards[card.keyword]
+                    assert card.value == new_card.value
```
Maybe one option would be to compare the coldefs attributes instead, with something like this (that could go in a new helper function):
```
from astropy.io.fits.column import ColumnAttribute

assert hdu.columns.names == new_tbhdu.columns.names
attrs = [k for k, v in fits.Column.__dict__.items() if isinstance(v, ColumnAttribute)]

for name in hdu.columns.names:
    col = hdu.columns[name]
    new_col = new_tbhdu.columns[name]
    for attr in attrs:
        if getattr(col, attr) and getattr(new_col, attr):
             assert getattr(col, attr) == getattr(new_col, attr)
```

cmarmo commented:
Thanks @saimn for your suggestion.
Do my changes correspond to what you had in mind?

cmarmo commented:
I don't see where the error on documentation comes from... should I synchronize with upstream?

pllim commented:
> should I synchronize with upstream?

Yes, please! Thanks.

saimn commented on astropy/io/fits/tests/test_convenience.py:
```diff
@@ -255,6 +271,51 @@ def test_tabledump(self):
+                    new_col = tbhdu.columns[name]
+                    for attr in attrs:
+                        if getattr(col, attr) and getattr(new_col, attr):
+                            assert getattr(col, attr) == getattr(new_col, attr)
```
To avoid the duplication, could you put the function in `test_table` and import it here ?

saimn reviewed (changes requested):
Thanks @cmarmo, seems almost good, just a little change to avoid duplicating the comparison function. Also could you please rebase on main to remove the merge commit ?

cmarmo commented:
> Also could you please rebase on main to remove the merge commit ?

Should I also squash the commits? Thanks!

saimn reviewed (approved):
Thanks @cmarmo !

cmarmo commented:
Thanks @saimn for your time and your patience.
